### PR TITLE
Guardian Weekly banner (for Aus)

### PR DESCRIPTION
**Pull request by damien-shaw-guardian**
## What does this change?
This PR adds a new subscription banner the Guardian Weekly banner to be displayed in Australia

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

## Screenshots
- Desktop
Screen Shot 2020-03-26 at 15 27 06

- Tablet
Screen Shot 2020-03-26 at 15 28 18

- Mobile
Screen Shot 2020-03-26 at 15 29 38

## What is the value of this and can you measure success?
This is drive subscription uptake in Australia 

## Checklist
- New Design
- New Tracking codes
- New Unit Tests

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [ ] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [x] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/master/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [x] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [x] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/subscription-banner.js [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                { feature: 'subscriptions-banner' },
+                false
+            );
+            return true;
+        });
+
+const threeOrMorePageViews = (currentPageViews: number) =>
+    currentPageViews >= 3;
+
+const pageIsIdentity = (): boolean => {
```
Comment (ripecosta) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/subscription-banner.js]: Would suggest changing this to:
```
const pageIsIdentity = (): boolean => config.get('page.contentType') === 'Identity' || config.get('page.section') === 'identity';
```

File: package.json [hunk 2]
```diff
@@ -162,7 +162,8 @@
     "setupFilesAfterEnv": [
       "<rootDir>/dev/jest.setupTestFrameworkScriptFile.js"
     ],
-    "testEnvironment": "jest-environment-jsdom-global"
+    "testEnvironment": "jest-environment-jsdom-global",
+    "testURL": "http://test-url.com"
```
Comment (ripecosta) [inline package.json]: Could you clarify why this change is needed? I don't understand it's impact in the contact of this PR.

File: static/src/javascripts/projects/commercial/modules/header-bidding/prebid/prebid.spec.js [hunk 3]
```diff
@@ -59,7 +59,7 @@ describe('initialise', () => {
             _disableAjaxTimeout: false,
             _mediaTypePriceGranularity: {},
             _priceGranularity: 'custom',
-            _publisherDomain: 'http://localhost',
+            _publisherDomain: 'http://test-url.com',
```
Comment (ripecosta) [inline static/src/javascripts/projects/commercial/modules/header-bidding/prebid/prebid.spec.js]: See comment above. Seems like this change was required because of the changes to `package.json` but it's unclear why that change is necessary.

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            <p>Already a subscriber?</p>
+            <br class="temp-mobile-break">
+            <a
+                id="js-site-message--subscription-banner__sign-in"
+                class="site-message--subscription-banner__subscriber-link"
+                data-link-name="subscription-banner : sign in"
+                href="${signInUrl}"
+
+            >
+                <span class="site-message--subscription-banner__sign-in-link">Sign in</span> to not see this again
```
Comment (ripecosta) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: The message "Sign in to not see this again" is not completely accurate since if a user logs off they'll see the message again. I'm wondering if a message like "Sign in to stop seeing this message" would be more clear. Question for UX really.

**Review by ripecosta — COMMENTED** [review]
Left a couple small comments but the `testURL` thing has intrigued me.
Everything else looks fine 👍

File: package.json (hunk 2 above)
Comment (damien-shaw-guardian) [reply package.json]: I added a test url to give jsdom a consistent url in the tests, before I added this I undefined as the url in the tests

Comment (ripecosta) [reply package.json]: On the prebid tests?

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+                mockTracking,
+                gwBannerTemplate,
+                false
+            );
+
+            show();
+        });
+
+        it('Should remove banner when close button is clicked', () => {
+            let appendedBanner = 'Banner Not Appended';
```
Comment (ripecosta) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js]: Can we avoid using `let` here?
Maybe like this:
```
const appendedBanner = document.body ? document.body.innerHTML : ''
```

Comment (ripecosta) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js]: Similar improvements can be done in the file `subscription-banner.spec.js` below.

File: package.json (hunk 2 above)
Comment (ripecosta) [reply package.json]: Spoke to Damian about this 'offline' (which is actually online because everyone's WFH) and we're reached a healthy compromise. Yay! 🎉

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+                gwBannerTemplate,
+                isUserSignedIn
+            );
+            show();
+
+            const signInContainer = document.querySelector(
+                '.site-message--subscription-banner__sign-in'
+            );
+
+            const hasSignedInClass: ?boolean =
```
Comment (ripecosta) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js]: You don't need to keep the boolean type here anymore since the expression following it will always return a boolean. But if you want to keep it you should remove the optional `?` symbol since this expression will always definitely return a boolean value. Same thing goes for your other test file.

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            </a>
+        </div>
+
+        <div class="site-message--packshot-container">
+
+            <picture>
+                <source srcset="https://media.guim.co.uk/8937356626c6e4dcdf0f14ccee395058f7c0ef87/36_267_1815_889/500.png" media="(max-width: 739px)">
+                <source srcset="https://media.guim.co.uk/18cf1b898a18ba9715be2fd731aa91b028b258e7/280_42_1322_1114/500.png" media="(max-width: 1139px)">
+                <source srcset="https://media.guim.co.uk/7552e031526eceae6385685876a4d52fbbf485e8/19_8_1814_1013/500.png" media="(min-width: 1140px)">
+                <img srcset="https://media.guim.co.uk/7552e031526eceae6385685876a4d52fbbf485e8/19_8_1814_1013/500.png" alt="…">
```
[the remark this answers is not in the document]
Comment (jfsoul) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: I can show you how to do this.

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/subscription-banner-tracking.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+export type BannerTracking = {
+    signInUrl: string,
+    gaTracking: () => any,
+    subscriptionUrl: string,
+    trackBannerView: () => void,
+    trackBannerClick: (button: any) => void,
+    trackCloseButtons: (button: any) => void,
+};
+
+const DISPLAY_EVENT_KEY = 'subscription-banner : display';
```
Comment (jfsoul) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/subscription-banner-tracking.js]: might want to consider pulling out the prefix `subscription-banner: ` since it is repeated a few times.

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/subscription-banner.js [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        tracking.trackCloseButtons
+    );
+
+    return Promise.resolve(true);
+};
+
+const chooseBanner = (region: ReaderRevenueRegion) =>
+    region === 'australia' ? gwBannerTemplate : subscripionBannerTemplate;
+const show = createBannerShow(
+    bannerTracking('australia'),
```
Comment (jfsoul) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/subscription-banner.js]: don't forget to revert this - presumably this is where we use the geolocation?

File: static/src/stylesheets/module/site-messages/_subscription-medium-banner.scss [hunk 10]
```diff
@@ -388,3 +388,126 @@
 }
 
 /*********************** footer hacks END *********************/
+
+/*********************** gw Banner *********************/
```
Comment (jfsoul) [inline static/src/stylesheets/module/site-messages/_subscription-medium-banner.scss]: in an ideal world all this would be unnecessary:
a) we'd be able to find a consistent layout that accommodated different copy, packshots
b) tooling would control basic things like background colour

One for the future.

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+    getReaderRevenueRegion: jest.fn().mockReturnValue('united-kingdom'),
+}));
+jest.mock('common/modules/identity/api.js', () => ({
+    isUserLoggedIn: jest.fn().mockReturnValue(true),
+}));
+jest.mock('lib/report-error', () => jest.fn());
+
+const removeAllSpaces = (str: string) => str.replace(/\s/g, '');
+
+describe('Guardian Weekly Banner', () => {
```
Comment (jfsoul) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js]: what's the value of testing this and the digital banner separately? and if there is value, then could we combine the 2 sets of tests and just run them twice, once for `gwBannerTemplate` and once for the other template

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js (hunk 6 above)
Comment (damien-shaw-guardian) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js]: this expression can now return null so I've kept the optional `?` symbol here

Comment (ripecosta) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js]: The `&&` symbol will automatically change the expression's result into a boolean, despite any items in the expressions returning null.

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js (hunk 11 above)
Comment (damien-shaw-guardian) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner.spec.js]: Because the banners can change independently, I though it would be wise to test both of them to protect against changes that are made to them at different times

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js [hunk 12]
```diff
@@ -0,0 +1,115 @@
+// @flow
+
+import closeCentralIcon from 'svgs/icon/close-large.svg';
+import theGuardianLogo from 'svgs/logo/the-guardian-logo.svg';
+
+const isUserLoggedIn = userLoggedIn =>
```
Comment (jfsoul) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: I think this should be renamed - it sounds like it returns a boolean at the moment

Comment (damien-shaw-guardian) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: it does return a boolean

Comment (jfsoul) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: `isUserLoggedIn` is a function with the signature `(boolean) => string`, so it returns a string unless I'm mistaken.

Comment (jfsoul) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: `loggedInClassModifier` or similar would be more suitable IMO

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+                <span class="site-message--subscription-banner__sign-in-link">Sign in</span> to not see this again
+            </a>
+        </div>
+
+        <div class="site-message--packshot-container">
+
+            <picture>
+                <source srcset="https://i.guim.co.uk/img/media/8937356626c6e4dcdf0f14ccee395058f7c0ef87/36_267_1815_889/1815.png?width=324&quality=85&s=fee17219cbe86ecce70b3d872c848aab" media="(max-width: 739px)">
+                <source srcset="https://i.guim.co.uk/img/media/18cf1b898a18ba9715be2fd731aa91b028b258e7/280_42_1322_1114/1322.png?width=344&quality=85&s=d5b95c838557590a4db5dc274ab706cc" media="(max-width: 1139px)">
+                <source srcset="https://i.guim.co.uk/img/media/7552e031526eceae6385685876a4d52fbbf485e8/19_8_1814_1013/1814.png?width=541&quality=85&s=a68601dd33793d2f9056f5e91c797289" media="(min-width: 1140px)">
```
Comment (jfsoul) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: I think using an exact width e.g. 541 is possibly overkill, but equally it doesn't do too much harm.

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    signInUrl: string,
+    userLoggedIn: boolean
+): string =>
+    `<div class="site-message js-site-message js-double-site-message site-message--banner site-message--double-banner"
+          id="js-site-message--subscription-banner__holder"
+          tabindex="-1"
+          role="dialog"
+          aria-label="welcome"
+          aria-describedby="site-message__message"
+          data-component="AcquisitionsEngagementBannerStylingTweaks_control"
```
Comment (jfsoul) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: just noticed this - I feel like this is a remnant of the contributions banner and probably isn't right for us?

File: static/src/stylesheets/module/site-messages/_subscription-medium-banner.scss [hunk 15]
```diff
@@ -390,3 +390,164 @@
...
+        @include mq($from: leftCol) {
+            width: 46%;
+        }
+
+        @include mq($from: wide) {
+            width: 47%;
+        }
+
+        p {
+            line-height: 1.4;
```
Comment (jfsoul) [inline static/src/stylesheets/module/site-messages/_subscription-medium-banner.scss]: Should we add the unit?

File: static/src/stylesheets/module/site-messages/_subscription-medium-banner.scss [hunk 16]
```diff
@@ -390,3 +390,164 @@
...
+                max-width: 325px;
+                margin-right: 0;
+                margin-left: auto;
+            }
+
+            @media (min-width: 781px) {
+                max-width: 320px;
+            }
+
+            @media (min-width: 889px) {
```
Comment (jfsoul) [inline static/src/stylesheets/module/site-messages/_subscription-medium-banner.scss]: Why do we have these custom breakpoints? They concern me a little bit because we really ought to be trying to move to a more consistent responsive design that can handle different packshots and copy - these make it seem a bit more fragile to change.

**Review by jfsoul — APPROVED** [review]
I don't think any of the remaining comments I've flagged should block releasing this - but feel free to either do those in a follow up PR or quickly fix now.

Nice work!

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js (hunk 12 above)
Comment (damien-shaw-guardian) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: this is the `isUserLoggedIn` function I'm using `const isUserLoggedIn = (): boolean => getUserFromCookie() !== null`

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js (hunk 13 above)
Comment (damien-shaw-guardian) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: yeah your right, I was just trying get an exact copy of the design

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js (hunk 12 above)
Comment (jfsoul) [reply static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: Yes that's the function you're calling in another file whose return value eventually gets put in the argument `userLoggedIn` here.

But the `const` function on line 6 of this file is called `isUserLoggedIn` and yet returns a string, either empty string `''` or `'site-message--subscription-banner__sign-in--already-signed-in'`.

I feel like we're getting our wires crossed, so leave this if you like and we can sort it later.

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+    userLoggedIn: boolean
+): string => `
+<div id="js-subscription-banner-site-message" class="site-message--subscription-banner site-message--gw-banner">
+    <div class="site-message--subscription-banner__inner">
+        <h3 class="site-message--subscription-banner__title">
+           Read The Guardian in Print
+        </h3>
+
+        <div class="site-message--subscription-banner__description">
+            <p>Support The Guardian's independent journalism by subscribing
```
Comment (jfsoul) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: I'm not sure most would notice but the apostrophe here should probably be a proper one i.e. ’  rather than '

File: static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+const subscriptionBannerTemplate = (
+    subscriptionUrl: string,
+    signInUrl: string,
+    userLoggedIn: boolean
+): string => `
+<div id="js-subscription-banner-site-message" class="site-message--subscription-banner site-message--gw-banner">
+    <div class="site-message--subscription-banner__inner">
+        <h3 class="site-message--subscription-banner__title">
+           Read The Guardian in Print
```
Comment (jfsoul) [inline static/src/javascripts/projects/common/modules/ui/subscription-banners/guardian-weekly-banner-template.js]: The crowd seem to think this should be a lowercase p for print

**Review by jfsoul — APPROVED** [review]
